#+OPTIONS: toc:nil
These notes correspond to TCRD major version 6.
* UniProt
** *Download Required:* Yes. Download and uncompress XML files from the following URLs:
   https://www.uniprot.org/uniprot/?query=reviewed:yes AND organism:"Homo sapiens (Human) [9606]"
   https://www.uniprot.org/uniprot/?query=organism:"Mus musculus (Mouse) [10090]"
   https://www.uniprot.org/uniprot/?query=organism:"Rattus norvegicus (Rat) [10116]"
** *Pre-processing required:* None
** *Manual post-processing:* Yes
   Some UniProt records have multiple Gene IDs. The loader takes the first
one, but this is not always the right one. So manual fixes to Gene IDs
are in the file SQL/update_geneids.sql
** *Approx. load time*: 30 minutes
* HGNC
** *Download Required:* Yes
   Download a custom TSV file from
   https://www.genenames.org/download/custom/ with fields: HGNC ID
   Approved symbol, Approved name,Status,UniProt ID, NCBI Gene ID, Mouse genome database ID
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: 30 minutes
* NCBI GI Numbers
** *Download Required:* No, done by loader
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: < 5 minutes
* Ensembl Gene IDs
** *Download Required:* Yes. This uses the same XML files used for UniProt load, plus download and uncompress the following TSV files:
   ftp://ftp.ensembl.org/pub/current_tsv/homo_sapiens/Homo_sapiens.GRCh38.94.uniprot.tsv.gz
   ftp://ftp.ensembl.org/pub/current_tsv/mus_musculus/Mus_musculus.GRCm38.94.uniprot.tsv.gz
   ftp://ftp.ensembl.org/pub/current_tsv/rattus_norvegicus/Rattus_norvegicus.Rnor_6.0.94.uniprot.tsv.gz
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: 5 minutes
* NCBI Gene
** *Download Required:* No. Data loaded via API.
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: 8 hours
* STRING IDs
** *Download Required:* Yes
   The following files, or updated versions thereof should be downloaded
   and uncompressed:
*** http://string-db.org/mapping_files/uniprot/human.uniprot_2_string.2018.tsv.gz
*** https://stringdb-static.org/download/protein.aliases.v11.0.txt.gz
** *Pre-processing required:* Yes
   The proteinaliases file should be filtered for human entries:
   $ grep '^9606.' protein.aliases.v11.0.txt > 9606.protein.aliases.v11.0.txt
** *Manual post-processing:* No
** *Approx. load time*: 10 minutes
* JensenLab PubMed Scores
** *Download Required:* No, done by loader
** *Pre-processing required:* None
** *Manual post-processing:* Yes, but technically optional.
   Generate and run SQL to set missing values to zero. See file doc/README_v5.txt.
** *Approx. load time*: 10 minutes

* Antibodypedia
** *Download Required:* No. Data loaded via API.
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: 1.5 days
* IDG Families
** *Download Required:* No
   The required file is included in the data/ directory.
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: < 5 minutes
* Drug Central
** *Download Required:* Yes.
   Four files are required. A link is provided on the Drug Central
   downlaod page (http://drugcentral.org/download). Look for "Files required for TCRD import. Download Files Here.".
*** tclin_MMDDYYYY.tsv
*** tchem_drugs_MMDDYYYY.tsv
*** drug_info_MMDDYYYY.tsv
*** drug_indications_MMDDYYYY.tsv
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: 15 minutes
* ChEMBL 
** *Download Required:* Yes
   Go here ftp://ftp.ebi.ac.uk/pub/databases/chembl/ChEMBLdb/latest and
   get the latest MySQL dump tarball. E.g. chembl_24_1_mysql.tar.gz
** *Pre-processing required:* Yes
   ChEMBL MySQL database must be built from download. Instructions are
   included in the targall.
** *Manual post-processing:* None
** *Approx. load time*: 30 minutes
* Guide to Pharmacology (ie. IUPHAR)
** *Download Required:* No, done by loader
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: < 5 minutes
* GO Experimental Leaf Term Flags
** *Download Required:* No, done by loader
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: 10 minutes
* TDLs
** *Download Required:* No
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: 10 minutes

* Disease Ontology
** *Download Required:* No, done by loader
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: < 5 minutes
* Mammalian Phenotype Ontology
** Download Required:* No, done by loader
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: < 5 minutes
* RGD Disease Ontology
** Download Required:* No, done by loader
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: < 5 minutes
* Uberon Ontology
** Download Required:* No, done by loader
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: < 5 minutes

* Orthologs
*** *Download Required:* No, done by loader
*** *Pre-processing required:* None
*** *Manual post-processing:* None
*** *Approx. load time*: 20 minutes
* HomoloGene
*** *Download Required:* No, done by loader
*** *Pre-processing required:* None
*** *Manual post-processing:* None
*** *Approx. load time*: 1 hour


* PubTator Text-mining Scores
** *Download Required:* No, done by loader
** *Pre-processing required:* Yes
   Run perl/pubtator_count.pl on the uncompressed download.
** *Manual post-processing:* None
** *Approx. load time*: 1 hour

* JensenLab DISEASES
*** *Download Required:* No, done by loader
*** *Pre-processing required:* None
*** *Manual post-processing:* None
*** *Approx. load time*: 40 minutes
* DisGeNET
*** *Download Required:* No, done by loader
*** *Pre-processing required:* None
*** *Manual post-processing:* None
*** *Approx. load time*: < 5 minutes
* Monarch Disease Associations
   This one is a little special. Data is loaded from a MySQL database on
   AWS created by Schurer lab. Contact John Turner <jpt55@med.miami.edu>
   for access.
*** *Download Required:* No
*** *Pre-processing required:* None
*** *Manual post-processing:* None
*** *Approx. load time*: 20 minutes
* Monarch Ortholog Disease Associations
   Also requires access to the Schurer lab MySQL database on AWS.
*** *Download Required:* No
*** *Pre-processing required:* None
*** *Manual post-processing:* None
*** *Approx. load time*: 1.5 hours
* Expression Atlas
*** *Download Required:* Yes
    ftp://ftp.ebi.ac.uk/pub/databases/microarray/data/atlas/experiments/atlas-latest-data.tar.gz
*** *Pre-processing required:* Yes
    Run R/exp-atlas_process.R on downloaded files.
*** *Manual post-processing:* None
*** *Approx. load time*: 20 minutes.
* CTD Disease Associations
*** *Download Required:* No, done by loader
*** *Pre-processing required:* None
*** *Manual post-processing:* None
*** *Approx. load time*: 5 minutes
* eRAM Disease Associations
*** *Download Required:* No
*** *Pre-processing required:* Yes
    Run python/scrape-eRAM.py to collect data.
*** *Manual post-processing:* None
*** *Approx. load time*: 5 minutes

* OMIM
** *Download Required:* No, done by loader
   BUT: One must register to get OMIM downloads. This gives a user-specific download link.
   NB. the phenotypic series file must be added to one's key's entitlements by OMIM staff.
   To view a list of all the data a key has access to, go to: https://omim.org/downloads/<key>
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: 5 minutes

* GWAS Catalog
** *Download Required:* Yes
   Go to https://www.ebi.ac.uk/gwas/docs/file-downloads and get the file with description "All associations with added ontology annotations".
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: 15 
* IMPC Phenotypes
** *Download Required:* Yes
   Download and uncompress the following, or latest versions, of:
   ftp://ftp.ebi.ac.uk/pub/databases/impc/release-9.2/csv/IMPC_genotype_phenotype.csv.gz
   ftp://ftp.ebi.ac.uk/pub/databases/impc/release-9.2/csv/IMPC_ALL_statistical_results.csv.gz
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: 3 hours
* JAX/MGI Mouse/Human Orthology Phenotypes
** *Download Required:* No, done by loader
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: < 5 minutes
* RGD Phenotypes
** *Download Required:* No, done by loader
** *Pre-processing required:* Yes
   The input file is produced by the Rscript R/process-RGD.R
   Details/explanation of that code are in notebooks/RGD.ipynb
** *Manual post-processing:* None
** *Approx. load time*: < 5 minutes

* IDG Eligible Targets List
** *Download Required:* Yes
   Download latest IDG Eligible list from GitHub repo.
** *Pre-processing required:* No
** *Manual post-processing:* None
** *Approx. load time*: < 5 minutes
* JensenLab TISSUES
** *Download Required:* No, one by loader
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: 6 hours
* GTEx
** *Download Required:* Yes
   Input file prepared by Jeremy Yang. See
   https://github.com/unmtransinfo/expression-profiles
   Also, a file containing a manually currated dict mapping tissue
   names to Uberon IDs. These are ones for which TCRDMP.get_uberon_id does not return a uid.
** *Pre-processing required:* Yes
   Input file prepared by Jeremy Yang. See https://github.com/unmtransinfo/expression-profiles
** *Manual post-processing:* None
** *Approx. load time*: 3 hours
* Human Protein Atlas
** *Download Required:* Yes
   http://www.proteinatlas.org/download/normal_tissue.tsv.zip
** *Pre-processing required:* Yes
   The input file is produced by the Rscript R/process-HPA.R
   Details/explanation of that code are in notebooks/HPA.ipynb
** *Manual post-processing:* None
** *Approx. load time*: 15 minutes
* Human Proteome Map
** *Download Required:* Yes
   Go to http://www.humanproteomemap.org/download.php, register and then download:
*** HPM_gene_level_epxression_matrix_Kim_et_al_052914.csv.gz
*** HPM_protein_level_expression_matrix_Kim_et_al_052914.csv.gz
** *Pre-processing required:* Yes
   Run R/hpm_gene.R and R/hpm_protein.R on the downloaded files.
** *Manual post-processing:* None
** *Approx. load time*: 3 hours
* Human Cell Atlas
** *Download Required:* Yes
   Supplementary data files from original paper.
*** File Table S1 from http://science.sciencemag.org/content/suppl/2017/05/10/science.aal3321.DC1
*** File Table S6 from http://science.sciencemag.org/content/suppl/2017/05/10/science.aal3321.DC1
** *Pre-processing required:* No
** *Manual post-processing:* None
** *Approx. load time*: 20 minutes
* Cell Surface Protein Atlas
** *Download Required:* Yes
   Supplementary data files from original paper.
*** S1_File.csv
** *Pre-processing required:* No
** *Manual post-processing:* None
** *Approx. load time*: < 5 minutes
* CCLE
** *Download Required:* Yes
   CCLE_DepMap_18q3_RNAseq_RPKM_20180718.gct
** *Pre-processing required:* Yes
   The input file is produced by the Rscript R/process-CCLE.R
Details/explanation of that code are in notebooks/CCLE.ipynb
** *Manual post-processing:* None
** *Approx. load time*: 3.5 hours
* Consensus Expression Values
*** *Download Required:* No
*** *Pre-processing required:* No
*** *Manual post-processing:* None
*** *Approx. load time*: 20 minutes





* JensenLab COMPARTMENTS
** *Download Required:* No
   Done by loader
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: 20 minutes


* KEGG Pathways
** *Download Required:* No
    Data loaded via API.
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: 30 minutes
* PathwayCommons
** *Download Required:* No, done by loader.
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: < 5 minutes
* Reactome Pathways
** *Download Required:* No, done by loader
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: < 5 minutes
* WikiPathways
** *Download Required:* No, done by loader
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: < 5 minutes
* TIN-X
** *Download Required:* No, done by pre-processing code
** *Pre-processing required:* Yes.
   Run python/TIN-X.py
** *Manual post-processing:* None
** *Approx. load time*: 10 hours

* EBI Patent Counts
** *Download Required:* Yes
   Data obtained directly from EBI.
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: < 5 minutes
* KEGG Distances
** *Download Required:* No
** *Pre-processing required:* No
** *Manual post-processing:* None
** *Approx. load time*: < 5 minutes
* KEGG Nearest Tclin
** *Download Required:* No
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: < 5 minutes
* LINCS L1000 XRefs
** *Download Required:* Yes
** *Pre-processing required:* None
** *Manual post-processing:* No
** *Approx. load time*: < 5 minutes





* LocSigDB
** *Download Required:* No, done by loader
** *Pre-processing required:* None
** *Manual post-processing:* No
** *Approx. load time*: 5 minutes

* PANTHER Classes
** *Download Required:* Yes
*** ftp://ftp.pantherdb.org/sequence_classifications/current_release/PANTHER_Sequence_Classification_files/PTHR14.1_human_
*** http://data.pantherdb.org/PANTHER14.1/ontology/Protein_Class_14.0
*** http://data.pantherdb.org/PANTHER14.1/ontology/Protein_class_relationship
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: < 5 minutes


* PubChem CIDs
  This must be (re-)run after DrugCentral or ChEMBL are (re-)loaded.
** *Download Required:* No, done by loader
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: 5 minutes







* Transcription Factor Flags
** *Download Required:* No, done by loader
** *Pre-processing required:* No
** *Manual post-processing:* None
** *Approx. load time*: < 5 minutes





* TMHMM Predictions
** *Download Required:* No
** *Pre-processing required:* Yes
   Go to http://www.cbs.dtu.dk/cgi-bin/nph-sw_request?tmhmm to download TMHMM software.
** *Manual post-processing:* None
** *Approx. load time*: 35 minutes
* PubMed Abstracts
  This must be (re-)run after TIN-X is loaded.
** *Download Required:* No
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: 40 hours
* GeneRIF Years
** *Download Required:* No
** *Pre-processing required:* Yes
   Run python/mk-PubMed2DateMap.py
** *Manual post-processing:* None
** *Approx. load time*: 20 minutes
* DRGC Resources
** *Download Required:* No
   Data loaded via RSS API
** *Pre-processing required:* No
** *Manual post-processing:* None
** *Approx. load time*: < 5 minutes


* Harmonizome
** *Download Required:* No
   Data loaded via API.
** *Pre-processing required:* No
** *Manual post-processing:* Yes
*** mysql> delete from gene_attribute_type where id not in (select distinct gat_id from gene_attribute);
*** Check log for GeneID mismatches and update accordingly
** *Approx. load time*: > 10 days
* Harmonogram CDFs
** *Download Required:* No
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: 2.5 hours








* LINCS
** *Download Required:* Yes
** *Pre-processing required:* Yes
** *Manual post-processing:* None
** *Approx. load time*: 

* IMPC Mice Clones
** *Download Required:* Yes
   Data obtained directly from IMPC.
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: < 5 minutes





* Reactome PPIs
** *Download Required:* No, done by loader.
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: < 5 minutes
* BioPlex PPIs
** *Download Required:* Yes
*** http://wren.hms.harvard.edu/bioplex/downloadInteractions.php
*** http://bioplex.hms.harvard.edu/data/BioPlex_interactionList_v4a.tsv
*** http://bioplex.hms.harvard.edu/data/interactome_update_MonYYYY.tsv
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: < 5 minutes
* STRINGDB
** *Download Required:* Yes
*** Download and uncompress:
    https://stringdb-static.org/download/protein.links.v11.0/9606.protein.links.v11.0.txt.gz
** *Pre-processing required:* None
** *Manual post-processing:* None
** *Approx. load time*: 2 hours

* DTO
** *Download Required:* Yes
   These files obtained from the Schurer Lab at UMiami.
*** For IDs and Classifications:
**** Final_ProteomeClassification_Sep232019.csv
**** DTO2UniProt_DTOv2.csv
*** For term hierarchy:
**** dto_proteome_classification_only.owl
** *Pre-processing required:* None
** *Manual post-processing:* Yes
*** Manual database updates for root nodes:
   mysql> UPDATE dto SET parent = NULL WHERE name IN ('GPCR', 'Kinase', 'Nuclear hormone receptor', 'Ion channel');
*** Then create foreign key:
    mysql> ALTER TABLE dto ADD CONSTRAINT fk_dto_dto FOREIGN KEY dto_idx1(parent) REFERENCES dto(id) ON DELETE RESTRICT;
** *Approx. load time*: < 5 minutes
* ClinVar
** *Download Required:* No, done by loader
** *Pre-processing required:* No
** *Manual post-processing:* None
** *Approx. load time*: 30 minutes
* IDG Eligible Target Lists
** *Download Required:* No
   List exported from previous TCRD versions.
** *Pre-processing required:* Yes
   List exported from previous TCRD versions.
** *Manual post-processing:* None
** *Approx. load time*: < 5 minutes

